File: ode/pendulum_on_tube/python/mp-lagrange.py
import numpy as np
import matplotlib.pyplot as plt
from scipy.optimize import bisect
from scipy.integrate import solve_ivp
from scipy.integrate import quad

import mpconfig as config

# constant of motion
def energy(x):
    theta, omega = x
    return 0.5*omega**2 + 1 - np.cos(theta)

# evolution in time
def pendulum(t, x):
    theta, omega = x
    d_theta = omega
    d_omega = -np.sin(theta)
    return [d_theta, d_omega]

# period of librations
e0 = energy(config.x_init)
print("E0 = ", e0)
amplitude = lambda theta: e0 - energy([theta, 0])
theta_0 = bisect(amplitude, 0, np.pi)
print("theta_0 = ", theta_0)
result = quad(lambda theta: 1/np.sqrt(amplitude(theta)), 0, theta_0)
t_period = 4*result[0]/np.sqrt(2)
print("T = ", t_period)

# ODE numerical solution
sol = solve_ivp(
    pendulum,
    [0, config.K_cycles*t_period],
    config.x_init,
    t_eval=np.arange(0, config.K_cycles*t_period, t_period/config.N_points),
    method="DOP853",
    rtol=1e-10,
    atol=1e-10
)

# plot phase orbit
fig = plt.figure(figsize=(6, 6*1.0))
plt.title("Mathematical pendulum. Lagrangian dynamics")
plt.axis("equal")
plt.xlabel(r"$\theta$")
plt.ylabel(r"$\omega$")
plt.grid(True)

plt.plot(
    sol.y[0],
    sol.y[1],
    fillstyle='none',
    marker='o',
    markersize=1.5,
    linestyle='None',
    alpha=0.5
)
xmin, xmax, ymin, ymax = plt.axis()
plt.text(
    xmax - 0.20*(xmax - xmin), ymin + 0.08*(ymax - ymin),
    r"$E_0$ = {:.4f}".format(e0) + "\n" + r"$T$ = {:.4f}".format(t_period)
)

plt.show()
fig.savefig('mp-lagrange-orbit.png', dpi=150)

# plot relative error in energy value
fig = plt.figure(figsize=(5*1.5, 5))
plt.title("Mathematical pendulum. Invariant error")
plt.xlabel("Time")
plt.ylabel("Relative energy error")
plt.grid(True)

plt.plot(sol.t, (energy(sol.y) - e0)/e0)
plt.legend(['Energy'])
xmin, xmax, ymin, ymax = plt.axis()
plt.text(
    xmin + 0.08*(xmax - xmin), ymin + 0.08*(ymax - ymin),
    r"$E_0$ = {:.4f}".format(e0) + "\n" + r"$T$ = {:.4f}".format(t_period)
)

plt.show()
fig.savefig('mp-lagrange-error.png', dpi=150)

File: ode/pendulum_on_tube/julia/mp-lagrange.jl
using DifferentialEquations
using Elliptic
using Plots
using Statistics

include("mp-config.jl")

function pendulum(du, u, p, t)
    du[1] = u[2]
    du[2] = -sin(u[1])
end

energy(u) = 0.5*u[2]^2 + 1 - cos(u[1])

e0 = energy(u0)
t_period = 4*Elliptic.K(0.5*e0)
println("energy = ", e0)
println("period = ", t_period)
t_step = t_period/N 
t_max = K*t_period

prob = ODEProblem(pendulum, u0, (0.0, t_max))
sol = solve(prob, Vern6(), adaptive=false, dt=t_step)
#display(sol)

gr(size = (300, 300))
Plots.scalefontsizes()
Plots.scalefontsizes(0.4)
plotOrbit = scatter(
    Tuple.(sol.u),
    aspect_ratio = 1,
    xlabel = "\\theta",
    ylabel = "\\omega",
    plot_title = "Mathematical pendulum - Lagrangian dynamics",
    markersize = 1.0,
    alpha = 0.25,
    legend = false,
    dpi = 300
)
annotate!(
    xlims(plotOrbit)[1] + 0.18*(xlims(plotOrbit)[2] - xlims(plotOrbit)[1]),
    ylims(plotOrbit)[2] - 0.06*(ylims(plotOrbit)[2] - ylims(plotOrbit)[1]),
    text("\$\\mathcal{E}_0 = $(round(e0, digits=4))\$\n\$T = $(round(t_period, digits=6))\$", :crimson, :right, 4)
)
#gui(), readline()
savefig(plotOrbit, "mp-lagrange-orbit.png")

constant_of_motion = map(energy, sol.u)
# display(constant_of_motion)
println("mean energy = ", mean(constant_of_motion))
println("std energy = ", std(constant_of_motion))

gr(size = (360, 240))
Plots.scalefontsizes()
Plots.scalefontsizes(0.4)
plotError = scatter(
    sol.t, (constant_of_motion .- energy(u0))./energy(u0),
    xlabel = "Time",
    ylabel = "Relative energy error",
    plot_title = "Mathematical pendulum - Lagrangian dynamics",
    markersize = 0.8,
    alpha = 0.15,
    legend = false,
    dpi = 300,
    annotation = (t_max - t_period, 0.0, text("\$\\mathcal{E}_0 = $(round(e0, digits=4))\$\n\$T = $(round(t_period, digits=6))\$", :darkslateblue, :bottom, 4))
)
# gui(), readline()
savefig(plotError, "mp-lagrange-error.png")
